editor/scene/gui/container_editor_plugin.cpp

```diff
@@ -0,0 +1,290 @@
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "container_editor_plugin.h"
+
+#include "core/math/math_funcs.h"
+#include "editor/scene/canvas_item_editor_plugin.h"
+#include "editor/themes/editor_scale.h"
+#include "scene/gui/aspect_ratio_container.h"
+#include "scene/gui/box_container.h"
+#include "scene/gui/center_container.h"
+#include "scene/gui/flow_container.h"
+#include "scene/gui/grid_container.h"
+#include "scene/gui/margin_container.h"
+#include "scene/gui/scroll_container.h"
+
+ContainerEditorPlugin::ContainerType ContainerEditorPlugin::_get_container_type(Object *p_object) const {
+	if (Object::cast_to<GridContainer>(p_object)) {
+		return CONTAINER_TYPE_GRID;
+	} else if (Object::cast_to<AspectRatioContainer>(p_object)) {
+		return CONTAINER_TYPE_ASPECT_RATIO;
+	} else if (Object::cast_to<CenterContainer>(p_object)) {
+		return CONTAINER_TYPE_CENTER;
+	} else if (Object::cast_to<ScrollContainer>(p_object)) {
+		return CONTAINER_TYPE_SCROLL;
+	} else if (Object::cast_to<MarginContainer>(p_object)) {
+		return CONTAINER_TYPE_MARGIN;
+	} else if (Object::cast_to<SplitContainer>(p_object)) {
+		return CONTAINER_TYPE_SPLIT;
+	} else if (Object::cast_to<FlowContainer>(p_object)) {
+		return CONTAINER_TYPE_FLOW;
+	} else if (Object::cast_to<BoxContainer>(p_object)) {
+		return CONTAINER_TYPE_BOX;
+	}
+	return CONTAINER_TYPE_NONE;
+}
+
+void ContainerEditorPlugin::edit(Object *p_object) {
+	if (container) {
+		container->disconnect(SNAME("draw"), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
+		container->disconnect(SNAME("sort_children"), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
+		container->disconnect(SceneStringName(minimum_size_changed), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
+	}
+
+	container = Object::cast_to<Container>(p_object);
+	type = _get_container_type(p_object);
+
+	if (container) {
+		container->connect(SNAME("draw"), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
+		container->connect(SNAME("sort_children"), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
+		// minimum_size_changed is required for BoxContainer because it indirectly resorts its children instead of calling queue_resort when vertical is changed.
```

Comment:
```suggestion
		// `minimum_size_changed` is required for BoxContainer, because it indirectly resorts its children instead of calling `queue_resort()` when `vertical` is changed.
```